File: src/relationships.rs
/// Directed cycle traversal on N nodes
pub struct DirectedCycle {
    pub size: usize,
}

impl DirectedCycle {
    pub fn new(size: usize) -> Self {
        DirectedCycle { size }
    }

    /// Single step: (position + step) mod size
    pub fn step(&self, position: usize, step: i32) -> usize {
        let s = self.size as i32;
        ((position as i32 + step).rem_euclid(s)) as usize
    }

    /// Follow a chain of steps from a starting position
    pub fn chain(&self, start: usize, step: i32, depth: usize) -> Vec<usize> {
        let mut result = vec![start];
        let mut current = start;
        for _ in 0..depth {
            current = self.step(current, step);
            result.push(current);
        }
        result
    }
}

/// Structural pair table: explicit pairings not based on geometry
pub struct PairTable {
    pairs: Vec<(usize, usize)>,
}

impl PairTable {
    pub fn new(pairs: Vec<(usize, usize)>) -> Self {
        PairTable { pairs }
    }

    /// Find the structural partner of a given position
    pub fn partner(&self, position: usize) -> Option<usize> {
        for &(a, b) in &self.pairs {
            if a == position {
                return Some(b);
            }
            if b == position {
                return Some(a);
            }
        }
        None
    }

    /// Check if two positions are structurally paired
    pub fn are_paired(&self, a: usize, b: usize) -> bool {
        self.pairs
            .iter()
            .any(|&(x, y)| (x == a && y == b) || (x == b && y == a))
    }
}

File: python/relationships.py
"""
Structural relationship primitives: directed cycles and pair tables.

Translates src/relationships.rs — non-geometric relationship structures.
"""


class DirectedCycle:
    """Directed cycle traversal on N nodes."""

    def __init__(self, size: int):
        self.size = size

    def step(self, position: int, step: int) -> int:
        """Single step: (position + step) mod size."""
        return (position + step) % self.size

    def chain(self, start: int, step: int, depth: int) -> list[int]:
        """Follow a chain of steps from a starting position."""
        result = [start]
        current = start
        for _ in range(depth):
            current = self.step(current, step)
            result.append(current)
        return result


class PairTable:
    """Structural pair table: explicit pairings not based on geometry."""

    def __init__(self, pairs: list[tuple[int, int]]):
        self.pairs = pairs

    def partner(self, position: int) -> int | None:
        """Find the structural partner of a given position."""
        for a, b in self.pairs:
            if a == position:
                return b
            if b == position:
                return a
        return None

    def are_paired(self, a: int, b: int) -> bool:
        """Check if two positions are structurally paired."""
        return any(
            (x == a and y == b) or (x == b and y == a)
            for x, y in self.pairs
        )
